services/powchain/src/main/java/tech/pegasys/teku/services/powchain/execution/ExecutionEngineService.java

```diff
@@ -57,45 +63,99 @@ public void chainHeadUpdated(
       Bytes32 currentDutyDependentRoot,
       Optional<ReorgContext> optionalReorgContext) {
 
-    recentChainData
-        .retrieveBlockByRoot(bestBlockRoot)
-        .thenCompose(
-            maybeABlock ->
-                maybeABlock
-                    .flatMap(block -> block.getBody().toVersionMerge())
-                    .map(this::onHeadBlockMerge)
-                    .orElseGet(() -> SafeFuture.completedFuture(null)))
-        .finish(err -> LOG.warn("setHead failed", err));
-  }
+    final Optional<Bytes32> curFinalizedBlock;
 
-  private SafeFuture<Void> onHeadBlockMerge(BeaconBlockBodyMerge blockBody) {
-    // Check if there is a payload
-    if (!blockBody.getExecution_payload().equals(new ExecutionPayload())) {
-      return executionEngineChannel.setHead(blockBody.getExecution_payload().getBlock_hash());
-    } else {
-      return SafeFuture.completedFuture(null);
+    synchronized (this) {
+      lastBestBlock = Optional.of(bestBlockRoot);
+      curFinalizedBlock = lastFinalizedBlock;
     }
+
+    updateForkChoice(bestBlockRoot, curFinalizedBlock);
   }
 
   @Override
   public void onNewFinalizedCheckpoint(Checkpoint checkpoint) {
-    recentChainData
-        .retrieveBlockByRoot(checkpoint.getRoot())
+    final Bytes32 curBestBlock;
+    final Optional<Bytes32> curFinalizedBlock;
+
+    synchronized (this) {
+      lastFinalizedBlock = Optional.of(checkpoint.getRoot());
+      if (lastBestBlock.isEmpty()) {
+        return;
+      }
+      curBestBlock = lastBestBlock.get();
+      curFinalizedBlock = lastFinalizedBlock;
+    }
+
+    updateForkChoice(curBestBlock, curFinalizedBlock);
+  }
+
+  private static Optional<Bytes32> getPowBlockHash(Optional<BeaconBlock> block) {
```

Comment:
We won't have PoW anymore, I'd suggest the following renaming:
```suggestion
  private static Optional<Bytes32> getPayloadBlockHash(Optional<BeaconBlock> block) {
```